politeiawww/mail/client.go

```diff
@@ -0,0 +1,237 @@
+// Copyright (c) 2020-2021 The Decred developers
+// Use of this source code is governed by an ISC
+// license that can be found in the LICENSE file.
+
+package mail
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"fmt"
+	"io/ioutil"
+	"net/mail"
+	"net/url"
+	"time"
+
+	"github.com/dajohi/goemail"
+	"github.com/decred/politeia/politeiawww/user"
+	"github.com/google/uuid"
+)
+
+// client provides an SMTP client for sending emails from a preset email
+// address.
+type client struct {
+	smtp        *goemail.SMTP        // SMTP server
+	mailName    string               // From name
+	mailAddress string               // From email address
+	disabled    bool                 // Has email been disabled
+	mailerDB    user.MailerDB        // User database in www
+	limit       int                  // Email rate limit
+	userEmails  map[string]uuid.UUID // User email to uuid
+}
+
+// IsEnabled returns whether the mail server is enabled.
+func (c *client) IsEnabled() bool {
+	return !c.disabled
+}
+
+// SendTo sends an email with the given subject and body to the provided list
+// of email addresses.
+func (c *client) SendTo(subject, body string, recipients []string) error {
+	if c.disabled || len(recipients) == 0 {
+		return nil
+	}
+
+	// Setup email
+	msg := goemail.NewMessage(c.mailAddress, subject, body)
+	msg.SetName(c.mailName)
+
+	// Add all recipients to BCC
+	for _, v := range recipients {
+		msg.AddBCC(v)
+	}
+
+	return c.smtp.Send(msg)
+}
+
+// SendToUsers sends an email with the given subject and body to the provided list
+// of email addresses. This adds an email rate limit functionality in order
+// to avoid spamming from malicious users.
+//
+// This function satisfies the Mailer interface.
+func (c *client) SendToUsers(subjects, body string, recipients []string) error {
+	valid, invalid, histories, err := c.filterRecipients(recipients)
+	if err != nil {
+		return err
+	}
+
+	// Handle valid recipients.
+	if len(valid) > 0 {
+		err := c.SendTo(subjects, body, valid)
+		if err != nil {
+			return err
+		}
+	}
+
+	// Handle invalid recipients.
+	if len(invalid) > 0 {
+		err = c.SendTo(limitEmailSubject, limitEmailBody, invalid)
+		if err != nil {
+			return err
+		}
+	}
+
+	// Update email histories on db.
+	if len(histories) > 0 {
+		c.mailerDB.EmailHistoriesSave(histories)
+	}
+
+	return nil
+}
+
+// filterRecipients divides recipients into valid, those that are able to
+// receive emails, and invalid, those that have hit the email rate limit,
+// but have not yet received the warning email. It also updates the email
+// history for each user inside the recipients list.
+func (c *client) filterRecipients(rs []string) ([]string, []string, map[uuid.UUID]user.EmailHistory, error) {
```

Comment:
This function is where all the rate limiting function resides.  Let's test it.  It will be easier now that we separated out the `MailerDB` into its own interface.  You'll need to create a `MailerDB` implemenation for testing that saves the email histories in memory.

Reply:
done 👍